### PR TITLE
Law: general solution of 1D wave equation

### PR DESCRIPTION
**Pull request by alesanter**

### CODE REVIEW

File: symplyphysics/laws/waves/wave_equation_general_solution_in_one_dimension.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+general_solution = symbols("general_solution", cls=SymFunction)
+left_traveling_solution = symbols("left_traveling_solution", cls=SymFunction)
+right_traveling_solution = symbols("right_traveling_solution", cls=SymFunction)
+angular_wavenumber = Symbol("angular_wavenumber", angle_type / units.length)
+position = Symbol("position", units.length)
+angular_frequency = Symbol("angular_frequency", angle_type / units.time)
+time = Symbol("time", units.time)
+
+law = Eq(
+    general_solution(time),
```
Comment (blackyblack) [inline symplyphysics/laws/waves/wave_equation_general_solution_in_one_dimension.py:36]: In description it is a function of position and time.

File: symplyphysics/laws/waves/wave_equation_general_solution_in_one_dimension.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    right_traveling_solution(angular_wavenumber * position - angular_frequency * time)
```
Comment (blackyblack) [inline symplyphysics/laws/waves/wave_equation_general_solution_in_one_dimension.py:37]: It is not clear, what are the parameters of this function. Requires additional description.

File: symplyphysics/laws/waves/wave_equation_general_solution_in_one_dimension.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    position_: Quantity,
+    angular_frequency_: Quantity,
+    time_: Quantity,
+    phase_lag_: Quantity | float,
+) -> Quantity:
+    """Solution of the wave equation in the form `A * cos(k*x + w*t + phi)`"""
+
+    solution = law.rhs.subs({
+        right_traveling_solution(angular_wavenumber * position - angular_frequency * time): (
+            amplitude_ * cos(angular_wavenumber * position - angular_frequency * time + scale_factor(phase_lag_))
```
Comment (blackyblack) [inline symplyphysics/laws/waves/wave_equation_general_solution_in_one_dimension.py:68]: Well, you've completely replaced the function. It seems, `k*x - w*t` cannot represent `A * cos(k*x + w*t + phi)`, so something is wrong with your function.
I suggest to use `right_traveling_solution(position, time)`, so it properly corresponds to replacement function.
However this law becomes very basic: `general_solution = solution_left + solution_right`

Comment (alesanter) [reply symplyphysics/laws/waves/wave_equation_general_solution_in_one_dimension.py:68]: I've though of a better solution which is basically to write $`u(x, t) = f(\psi(x, t))`$ and note that $`\psi(x, t)`$ is referring to the wave phase described in [another law](https://github.com/blackyblack/symplyphysics/blob/master/symplyphysics/laws/waves/phase_of_traveling_wave.py)

Comment (alesanter) [reply symplyphysics/laws/waves/wave_equation_general_solution_in_one_dimension.py:68]: The point is to show that the solution of the wave equation is any function of one variable whose expression contains the $`x`$ and $`t`$ variables only in the combination $`k x \pm \omega t`$. I suppose the new approach suffices to show that.

Comment (blackyblack) [reply symplyphysics/laws/waves/wave_equation_general_solution_in_one_dimension.py:68]: Sounds good.

File: symplyphysics/laws/waves/wave_equation_general_solution_in_one_dimension.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+@validate_input(wave_phase_=wave_phase)
+@validate_output_same("amplitude_")
+def calculate_displacement(
+    amplitude_: Quantity,
+    wave_phase_: Quantity | float,
+) -> Quantity:
+    """Solution of the wave equation in the form `A * cos(psi)`"""
+    wave_phase_ = scale_factor(wave_phase_)
+    result = amplitude_ * law.rhs.subs(solution_function(wave_phase(position, time)), cos(wave_phase_))
```
Comment (blackyblack) [inline symplyphysics/laws/waves/wave_equation_general_solution_in_one_dimension.py:50]: Why not `amplitude_ * cos(wave_phase_)` ?

**Review by blackyblack — APPROVED** [review]

File: symplyphysics/laws/waves/wave_equation_general_solution_in_one_dimension.py (hunk 4 above)
Comment (alesanter) [reply symplyphysics/laws/waves/wave_equation_general_solution_in_one_dimension.py:50]: I thought we usually substitute something in the law to get the result like I did here, but if it's fine to just plug in the formula directly, I'm okay with it

Comment (blackyblack) [reply symplyphysics/laws/waves/wave_equation_general_solution_in_one_dimension.py]: Lol, no. I meant `result = law.rhs.subs(solution_function(wave_phase(position, time)), amplitude_ * cos(wave_phase_))`

Comment (alesanter) [reply symplyphysics/laws/waves/wave_equation_general_solution_in_one_dimension.py]: Oh now I see hahah

**Review by blackyblack — APPROVED** [review]

**Pull request merged**